### PR TITLE
Extract yanked versions into a SkyFunction

### PR DESCRIPTION
**Pull request by fmeum**
Ensures that yanked version information is only downloaded once per module name, which reduces the number of files fetched during module resolution. On my machine, this reduces `bazel mod deps` runtime on Bazel itself by ~15%.

This also prepares for storing yanked version information in the lockfile.

Work towards https://github.com/bazelbuild/bazel/issues/20369

### CODE REVIEW

**Comment by satyanandak** [comment]
@fmeum  Could you please take a look at the failing checks?

**Comment by fmeum** [comment]
@satyanandak Should be fixed now.

**Review by meteorcloudy — APPROVED** [review]
Nice!

**Comment by fmeum** [comment]
@bazel-io fork 7.2.0

File: src/main/java/com/google/devtools/build/lib/bazel/bzlmod/YankedVersionsFunction.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+      // fetched.
+      return YankedVersionsValue.create(Optional.empty());
+    } catch (URISyntaxException e) {
+      // This should never happen since we obtain the registry URL from an already constructed
+      // registry.
+      throw new IllegalStateException(e);
+    }
+  }
+
+  static final class YankedVersionsException extends SkyFunctionException {
```
Comment (Wyverald) [inline src/main/java/com/google/devtools/build/lib/bazel/bzlmod/YankedVersionsFunction.java:72]: this doesn't seem to be used anywhere. could you remove it?

**Review by Wyverald — APPROVED** [review]

**Comment by fmeum** [comment]
@bazel-io fork 7.2.0

**Comment by iancha1992** [comment]
The changes in this PR have been included in [Bazel 7.2.0 RC1](https://github.com/bazelbuild/bazel/releases/tag/7.2.0rc1). Please test out the release candidate and report any issues as soon as possible.
If you're using Bazelisk, you can point to the latest RC by setting USE_BAZEL_VERSION=7.2.0rc1. Thanks!

**Pull request closed without merging**